File: src/lib/api/performance.ts
/** Realized trading performance: a day, a month, and the calendar grid. */

import { apiGet, type RequestOptions } from "@/lib/api/client";
import {
  toDailyTargetStatus,
  type DataSource,
  type WireCalendar,
  type WireDailyPerformance,
  type WireMonthlyPerformance,
} from "@/lib/api/types";
import type {
  DailyTargetStatus,
  DailyTradingPerformance,
  PerformanceTotals,
} from "@/types/trading";

/** A session plus the target-relative figures the backend already computed. */
export interface DailyPerformanceResult extends DailyTradingPerformance {
  dailyTarget: number;
  remainingToTarget: number;
  progressPercentage: number;
  status: DailyTargetStatus;
  winRate: number;
  source: DataSource;
}

export interface MonthlyPerformanceResult {
  year: number;
  month: number;
  monthlyTarget: number;
  remainingToTarget: number;
  progressPercentage: number;
  totals: PerformanceTotals;
  source: DataSource;
}

/** One calendar cell: the session plus its status under the current target. */
export interface CalendarDayResult extends DailyTradingPerformance {
  dailyTarget: number;
  status: DailyTargetStatus;
}

export interface CalendarResult {
  year: number;
  month: number;
  days: CalendarDayResult[];
  source: DataSource;
}

function toDaily(wire: WireDailyPerformance): DailyPerformanceResult {
  return {
    date: wire.date,
    realizedPnl: wire.realized_pnl,
    trades: wire.trades,
    wins: wire.wins,
    losses: wire.losses,
    dailyTarget: wire.daily_target,
    remainingToTarget: wire.remaining_to_target,
    progressPercentage: wire.progress_percentage,
    status: toDailyTargetStatus(wire.status),
    winRate: wire.win_rate,
    source: wire.source,
  };
}

export async function fetchTodayPerformance(
  options?: RequestOptions,
): Promise<DailyPerformanceResult> {
  return toDaily(
    await apiGet<WireDailyPerformance>("/performance/today", options),
  );
}

/**
 * Fetches one dated session.
 *
 * A date with no recorded session is a 404 by design — an absent record is a
 * miss, not a zero — so callers must decide whether that means "no trading
 * that day" or "data unavailable".
 */
export async function fetchPerformanceForDate(
  dateKey: string,
  options?: RequestOptions,
): Promise<DailyPerformanceResult> {
  return toDaily(
    await apiGet<WireDailyPerformance>(`/performance/${dateKey}`, options),
  );
}

export async function fetchMonthlyPerformance(
  period?: { year: number; month: number },
  options?: RequestOptions,
): Promise<MonthlyPerformanceResult> {
  const wire = await apiGet<WireMonthlyPerformance>(
    "/performance/monthly",
    options,
    period ? { year: period.year, month: period.month } : undefined,
  );

  return {
    year: wire.year,
    month: wire.month,
    monthlyTarget: wire.monthly_target,
    remainingToTarget: wire.remaining_to_target,
    progressPercentage: wire.progress_percentage,
    totals: {
      realizedPnl: wire.realized_pnl,
      trades: wire.trades,
      wins: wire.wins,
      losses: wire.losses,
      activeSessions: wire.active_sessions,
    },
    source: wire.source,
  };
}

export async function fetchCalendar(
  year: number,
  month: number,
  options?: RequestOptions,
): Promise<CalendarResult> {
  const wire = await apiGet<WireCalendar>("/performance/calendar", options, {
    year,
    month,
  });

  return {
    year: wire.year,
    month: wire.month,
    source: wire.source,
    days: wire.days.map((day) => ({
      date: day.date,
      realizedPnl: day.realized_pnl,
      trades: day.trades,
      wins: day.wins,
      losses: day.losses,
      dailyTarget: day.daily_target,
      status: toDailyTargetStatus(day.status),
    })),
  };
}

File: backend/app/schemas/performance.py
"""Performance API schemas."""

from __future__ import annotations

from datetime import date

from pydantic import Field

from app.domain.enums import DailyTargetStatus
from app.schemas.common import (
    ApiModel,
    CurrencyMixin,
    Money,
    NonNegativeMoney,
    Percentage,
    SourcedMixin,
)


class DailyPerformanceResponse(ApiModel, CurrencyMixin, SourcedMixin):
    """One session's realized outcome measured against its daily target.

    `progress_percentage` is intentionally uncapped: beating the target is
    information worth reporting. Clamping belongs to the progress bar, not to
    the number.
    """

    date: date
    realized_pnl: Money
    daily_target: int = Field(ge=0)
    remaining_to_target: NonNegativeMoney
    progress_percentage: Percentage
    status: DailyTargetStatus
    trades: int = Field(ge=0)
    wins: int = Field(ge=0)
    losses: int = Field(ge=0)
    win_rate: Percentage


class MonthlyPerformanceResponse(ApiModel, CurrencyMixin, SourcedMixin):
    """Month-to-date realized performance against the monthly goal."""

    year: int = Field(ge=1970, le=2200)
    month: int = Field(ge=1, le=12)
    realized_pnl: Money
    monthly_target: NonNegativeMoney
    remaining_to_target: NonNegativeMoney
    progress_percentage: Percentage
    trades: int = Field(ge=0)
    wins: int = Field(ge=0)
    losses: int = Field(ge=0)
    active_sessions: int = Field(ge=0)


class CalendarDayResponse(ApiModel):
    """A single recorded session as the calendar consumes it.

    Leaner than `DailyPerformanceResponse` on purpose: the calendar renders a
    grid of markers and needs the status plus the P&L, not the full derived
    breakdown. Keeping this payload small is what lets the whole month load in
    one request instead of one request per cell.
    """

    date: date
    realized_pnl: Money
    trades: int = Field(ge=0)
    wins: int = Field(ge=0)
    losses: int = Field(ge=0)
    daily_target: int = Field(ge=0)
    status: DailyTargetStatus


class PerformanceCalendarResponse(ApiModel, CurrencyMixin, SourcedMixin):
    """A month of recorded sessions.

    Only dates with a journal entry appear in `days`. A missing date means no
    session was recorded — the calendar renders that as a blank cell rather
    than a zero, because "no data" and "traded to flat" are different facts.
    """

    year: int = Field(ge=1970, le=2200)
    month: int = Field(ge=1, le=12)
    days: list[CalendarDayResponse]
